### PR TITLE
RUM Demo iOS App

### PR DESCRIPTION
**Pull request by akashp-splunk**
- Demonstrates the initialisation of RUM SDK for iOS app.
- Demonstrates how to initialise SplunkOtelCrashReporting SDK to start listening to crashes.
- Demonstrates how to create the custom events to log on Splunk Observability Cloud.
- Demonstrates how to create error events to log on Splunk Observability Cloud.
- Demonstrates the way to create custom Spans and nesting of Spans.

### CODE REVIEW

File: rum-demo-app-ios/RUM Mobile Demo App/Common/Html/sample1.html [hunk 1]
```diff
@@ -0,0 +1,312 @@
+<!DOCTYPE html><html><head>
```
Comment (seemk) [inline rum-demo-app-ios/RUM Mobile Demo App/Common/Html/sample1.html:1]: How are these html files generated? Or where do they come from?

**Review by seemk — COMMENTED** [review]
Can we get rid of the unnecessary files (e.g. `.DS_Store`) please?

File: rum-demo-app-ios/RUM Mobile Demo App/Common/Html/sample1.html (hunk 1 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/Common/Html/sample1.html:1]: sample1.html & sample2.html - These two files were used to load inside inapp webview. 
We updated it and current code is not using these two files. So I am removing these two files from the code.

**Comment by akashp-splunk** [comment]
> Can we get rid of the unnecessary files (e.g. `.DS_Store`) please?

Sure. I will remove that.

File: rum-demo-app-ios/RUM Mobile Demo App/APIService/DataService.swift [hunk 2]
```diff
@@ -31,6 +31,7 @@ class DataService{
             guard let url = URL.init(string: urlString) else {return}
             var request = URLRequest.init(url: url)
             request.timeoutInterval = 60.0
+            //        request.setValue("keep-alive", forHTTPHeaderField: "Connection")
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/APIService/DataService.swift:34]: @akashCrest Can we remove this commented line?

File: rum-demo-app-ios/README.md [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+// MARK: - Splunk RUM BeaconURL
+let beaconURL = "ENTER_YOUR_RUM_AUTH_TOKEN"
+    
+// MARK: - Splunk RUMAuth key
+let rumAuth = "ENTER_YOUR_RUM_AUTH_TOKEN"
+    
+// MARK: - Splunk tracer name
+let RUM_TRACER_NAME = "SplunkRum"
+    
+// MARK: - root URL
```
Comment (seemk) [inline rum-demo-app-ios/README.md:32]: Do we need the MARK remarks?

File: rum-demo-app-ios/README.md [hunk 4]
```diff
@@ -0,0 +1,207 @@
...
+
+Code Reference: [here](../main/RUM%20Mobile%20Demo%20App/Common/AppDelegate.swift#L55)
+
+## Set Global Attributes for RUM
+We can set the global attributes for RUM SDK. The code snippet below demonstrate the way to set location values in the global attributes of RUM SDK.
+
+```
+SplunkRum.setGlobalAttributes([
+                                "_sf_geo_lat": <Location Latitude>,
+                                "_sf_geo_long": <Location Longitude>
```
Comment (seemk) [inline rum-demo-app-ios/README.md:108]: Just an example, but I think we use `location.lat` and `location.lon` elsewhere

File: rum-demo-app-ios/RUM Mobile Demo App/APIService/BaseRequestData.swift [hunk 5]
```diff
@@ -0,0 +1,36 @@
+//
+//  BaseRequestData.swift
+//  RUM Mobile Demo App
+//
+//  Created by Akash Patel on 06/01/22.
+//
```
Comment (seemk) [inline rum-demo-app-ios/RUM Mobile Demo App/APIService/BaseRequestData.swift:6]: For this and the following files, do we need these comments?

File: rum-demo-app-ios/RUM Mobile Demo App/Common/PCLBlurEffectAlert/PCLBlurEffectAlert+Action.swift [hunk 6]
```diff
@@ -0,0 +1,48 @@
+//
+//  PCLBlurEffectAlert+Action.swift
+//  PCLBlurEffectAlert
+//  RUM Mobile Demo App
+//
+//  Created by Akash Patel on 06/01/22.
```
Comment (seemk) [inline rum-demo-app-ios/RUM Mobile Demo App/Common/PCLBlurEffectAlert/PCLBlurEffectAlert+Action.swift:6]: If this is from https://github.com/hryk224/PCLBlurEffectAlert I think we need better attribution here.

File: rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/CheckOutVC.swift [hunk 7]
```diff
@@ -0,0 +1,282 @@
...
+         if(arrOfCharacters.count > 0) {
+             for i in 0...arrOfCharacters.count-1 {
+                 modifiedCreditCardString.append(arrOfCharacters[i])
+                 if((i+1) % 4 == 0 && i+1 != arrOfCharacters.count){
+                     modifiedCreditCardString.append("-")
+                 }
+             }
+         }
+         return modifiedCreditCardString
+     }
```
Comment (seemk) [inline rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/CheckOutVC.swift:89]: This seems to be copypasted from stackoverflow, do we give proper credit in such cases?

File: rum-demo-app-ios/RUM Mobile Demo App/APIService/BaseRequestData.swift (hunk 5 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/APIService/BaseRequestData.swift:6]: This block of comment is autogenerated by Xcode whenever we create a swift file or header files. 
We can remove these line manually. 
Let me know if need to remove from all places.

File: rum-demo-app-ios/RUM Mobile Demo App/Assets/Local HTMLs/index.html [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+            <link rel="stylesheet" href="stylesheet.css">
+                <script src="https://cdn.signalfx.com/o11y-gdi-rum/latest/splunk-otel-web.js"
+                    crossorigin="anonymous"></script>
+                <script type="text/javascript">
+                    function initialiseRUM(data) {
+                        var jsons = JSON.parse(data)
+                        SplunkRum.init({
+                            beaconUrl: 'https://rum-ingest.' + jsons.realm  + '.signalfx.com/v1/rum',
+                            rumAuth: jsons.token,
+                            app: "iOS WebView Spectacular.com"
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/Assets/Local HTMLs/index.html:16]: @akashCrest Let's rename the app here as "RUM Demo App Webview"

Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/Assets/Local HTMLs/index.html:16]: @akashCrest Let's rename the app here as "RUM Demo App Webview"

File: rum-demo-app-ios/RUM Mobile Demo App/Common/AppDelegate.swift [hunk 9]
```diff
@@ -0,0 +1,189 @@
...
+                locationObj.startUpdatingLocation()
+            }
+        }
+                
+    
+        SplunkRum.initialize(beaconUrl: AppVariables.current.beaconURL , rumAuth: AppVariables.current.token ,options: SplunkRumOptions(debug: true, environment: AppVariables.current.environment))
+            
+            SplunkRumCrashReporting.start()
+            SplunkRum.setGlobalAttributes(["DeviceID": UIDevice.current.identifierForVendor?.uuidString as Any])
+            //https://ingest.us1.signalfx.com  -- realm URL
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/Common/AppDelegate.swift:59]: @akashCrest Let's remove the commented block of code

File: rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift [hunk 10]
```diff
@@ -0,0 +1,588 @@
...
+            lblTitle.backgroundColor = .white
+            viewForMenu.addSubview(lblTitle)
+            viewForMenu.frame.origin.x = UIScreen.main.bounds.width - (lblTitle.frame.width + 20)
+            viewForMenu.frame.origin.y = (window?.safeAreaInsets.top ?? 0) + 15
+            viewForMenu.frame.size.height = 45
+            viewForMenu.backgroundColor = .white
+            
+            
+            
+    //        let tapGesture = UITapGestureRecognizer.init(target: self, action: #selector(changeConfigURLButtonClicked(_:)))
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift:293]: @akashCrest Can we remove this commented block of code?

Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift:293]: @akashCrest Can we remove this commented block of code?

File: rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift [hunk 11]
```diff
@@ -0,0 +1,588 @@
...
+            
+            viewMainContainer.tag = 102010
+    //        viewMainContainer.bringSubviewToFront(viewForMenu)
+            
+            window?.addSubview(viewMainContainer)
+        }
+    }
+    
+    @objc func changeConfigURLButtonClicked(_ sender : UIButton) {
+        print("CLICKED BUTTON ++ ==>")
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift:325]: @akashp-splunk Can we remove this print statements?

Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift:325]: @akashp-splunk Can we remove this print statements?

File: rum-demo-app-ios/RUM Mobile Demo App/View/CollectionViewCell/CollectionReusableView.swift [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+    }
+    
+    func pickerView(_ pickerView: UIPickerView, viewForRow row: Int, forComponent component: Int, reusing view: UIView?) -> UIView {
+        let lblTitle = UILabel.init()
+        lblTitle.text = self.quantityArray[row]
+        lblTitle.textColor = .black
+        lblTitle.textAlignment = .center
+        lblTitle.isAccessibilityElement = true
+        lblTitle.accessibilityLabel = "product_detail_quantity_\(quantityArray[row])"
+        print((self.lblProductName.text ?? ""))
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/View/CollectionViewCell/CollectionReusableView.swift:65]: @akashCrest Can we remove the print statements?

Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/View/CollectionViewCell/CollectionReusableView.swift:65]: @akashCrest Can we remove the print statements?

File: rum-demo-app-ios/RUM Mobile Demo App/ViewModel/StaticEventsVM.swift [hunk 13]
```diff
@@ -0,0 +1,109 @@
...
+                                       "quantity" : 1]
+        DataService.request( URL , method: "POST", params:parameters, type: StaticEvents.self) { (staticevent, error , responsecode) in
+                self.responsecode = responsecode
+                self.staticevent = staticevent
+                self.error = error as? Error
+                completion?()
+        }
+    }
+    
+    fileprivate func generate4xxEndpoitError(completion: (()->Void)? = nil) {
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/ViewModel/StaticEventsVM.swift:57]: @akashCrest Can we fix the typo here?

Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/ViewModel/StaticEventsVM.swift:57]: @akashCrest Can we fix the typo here?

File: rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift [hunk 14]
```diff
@@ -0,0 +1,116 @@
...
+        btnFourHundredError.addTextSpacing()
+        btnFiveHundredError.addTextSpacing()
+        btnException.addTextSpacing()
+        btnSlowApiResponse.addTextSpacing()
+        btnLocalWebView.addTextSpacing()
+        btnShopWebView.addTextSpacing()
+    }
+    
+    // MARK: - Networking
+    private func attemptNetwrokCall(withcode : Int) {
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift:51]: @akashCrest Can we fix the typo here?

Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift:51]: @akashCrest Can we fix the typo here?

File: rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift [hunk 15]
```diff
@@ -0,0 +1,116 @@
...
+        viewModel.staticEvent(withcode: withcode)
+    }
+    
+    //MARK: - Actions
+     @IBAction func btnCrashAppClicked(_ sender: Any) {
+         viewModel.crashApp()
+    }
+
+    @IBAction func btnFourHundredErrorClicked(_ sender: Any) {
+        attemptNetwrokCall(withcode: 400)
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift:61]: @akashCrest Can we fix the typo here?

Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift:61]: @akashCrest Can we fix the typo here?

File: rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift [hunk 16]
```diff
@@ -0,0 +1,116 @@
...
+    }
+    
+    @IBAction func btnFiveHundredErrorClicked(_ sender: Any) {
+        attemptNetwrokCall(withcode: 500)
```
Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift:65]: @akashCrest Can we fix the typo here?

Comment (poojamehta-crest) [inline rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift:65]: @akashCrest Can we fix the typo here?

File: rum-demo-app-ios/RUM Mobile Demo App/Common/PCLBlurEffectAlert/PCLBlurEffectAlert+Action.swift (hunk 6 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/Common/PCLBlurEffectAlert/PCLBlurEffectAlert+Action.swift]: This dependency has been removed in latest code commit.

File: rum-demo-app-ios/RUM Mobile Demo App/Assets/Local HTMLs/index.html (hunk 8 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/Assets/Local HTMLs/index.html]: Done.

Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/Assets/Local HTMLs/index.html]: Done

File: rum-demo-app-ios/RUM Mobile Demo App/Common/AppDelegate.swift (hunk 9 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/Common/AppDelegate.swift]: Removed commented block.

File: rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift (hunk 10 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift]: Done

Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift]: Done

File: rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift (hunk 11 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift]: Print statement removed.

Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/Extensions/UIViewControllerExtension.swift]: Done

File: rum-demo-app-ios/RUM Mobile Demo App/View/CollectionViewCell/CollectionReusableView.swift (hunk 12 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/View/CollectionViewCell/CollectionReusableView.swift]: Done

Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/View/CollectionViewCell/CollectionReusableView.swift]: Done

File: rum-demo-app-ios/RUM Mobile Demo App/ViewModel/StaticEventsVM.swift (hunk 13 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/ViewModel/StaticEventsVM.swift]: Typo fixed

Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/ViewModel/StaticEventsVM.swift]: Typo fixed

File: rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift (hunk 14 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift]: Typo fixed

Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift]: Typo fixed.

File: rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift (hunk 15 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift]: Done

Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift]: Handled.

File: rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift (hunk 16 above)
Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift]: Done

Comment (akashp-splunk) [reply rum-demo-app-ios/RUM Mobile Demo App/View/ViewControllers/StaticEventsVC.swift]: Typo fixed.

**Pull request closed without merging**